modules/xfeatures2d/src/beblid.cpp

```diff
@@ -0,0 +1,602 @@
+    A = ptr[idx1];
+    B = ptr[idx2];
+    C = ptr[idx3];
+    D = ptr[idx4];
+
+    // Calculate the mean intensity value of the pixels in the box
+    sum1 = float(A + D - B - C);
+    box_area1 = (box1y2 - box1y1) * (box1x2 - box1x1);
+    CV_DbgAssert(box_area1 > 0);
+    average1 = sum1 / box_area1;
+
+    // Calculate the indices on the integral image where the box falls
+    idx1 = box2y1 * width + box2x1;
+    idx2 = box2y1 * width + box2x2;
+    idx3 = box2y2 * width + box2x1;
+    idx4 = box2y2 * width + box2x2;
+
+    CV_DbgAssert(idx1 >= 0 && idx1 < integralImage.size().area());
+    CV_DbgAssert(idx2 >= 0 && idx2 < integralImage.size().area());
+    CV_DbgAssert(idx3 >= 0 && idx3 < integralImage.size().area());
+    CV_DbgAssert(idx4 >= 0 && idx4 < integralImage.size().area());
+
+    // Read the integral image values for the first box
+    A = ptr[idx1];
+    B = ptr[idx2];
+    C = ptr[idx3];
+    D = ptr[idx4];
+
+    // Calculate the mean intensity value of the pixels in the box
+    sum2 = float(A + D - B - C);
+    box_area2 = (box2y2 - box2y1) * (box2x2 - box2x1);
+    CV_DbgAssert(box_area2 > 0);
+    average2 = sum2 / box_area2;
+
+    return average1 - average2;
+}
+
+// descriptor computation using keypoints
+void BEBLID_Impl::compute(InputArray _image, vector<KeyPoint> &keypoints, OutputArray _descriptors)
+{
+    Mat image = _image.getMat();
+
+    if (image.empty())
+        return;
+
+    if (keypoints.empty())
+        return;
+
+    Mat grayImage;
+    switch (image.type()) {
+    case CV_8UC1:
+        grayImage = image;
+        break;
+    case CV_8UC3:
+        cvtColor(image, grayImage, COLOR_BGR2GRAY);
+        break;
+    case CV_8UC4:
+        cvtColor(image, grayImage, COLOR_BGRA2GRAY);
+        break;
+    default:
+        CV_Error(Error::StsBadArg, "Image should be 8UC1, 8UC3 or 8UC4");
+    }
+
+    cv::Mat integralImg;
+
+    // compute the integral image
+    cv::integral(grayImage, integralImg);
+
+    // Create the output array of descriptors
+    _descriptors.create((int)keypoints.size(), descriptorSize(), descriptorType());
+
+    // descriptor storage
+    cv::Mat descriptors = _descriptors.getMat();
+    CV_DbgAssert(descriptors.type() == CV_8UC1);
+
+    // Compute the BEBLID descriptors
+    computeBEBLID(integralImg, keypoints, descriptors);
+}
+
+// constructor
+BEBLID_Impl::BEBLID_Impl(float scale_factor, int n_bits)
+    : scale_factor_(scale_factor), patch_size_(32, 32)
+{
+    // Populate wl_params_
+    if (n_bits == SIZE_512_BITS)
+    {
+        // Pre-trained parameters of BEBLID-512 trained in Liberty data set with
+        // a million of patch pairs, 20% positives and 80% negatives
+        wl_params_ =
+            {{24, 18, 15, 17, 6, 13}, {19, 14, 13, 17, 2, 18}, {23, 19, 12, 15, 6, 19}, {24, 14, 16, 16, 6, 11},
+             {16, 15, 12, 16, 1, 12}, {16, 15, 7, 10, 4, 10}, {17, 12, 8, 17, 3, 16}, {24, 12, 11, 17, 7, 19},
```

Comment:
Please move these blobs into dedicated files (to keep source code clean):
- `beblid.p512.hpp`
```
// Pre-trained parameters of BEBLID-512 trained in Liberty data set with
// a million of patch pairs, 20% positives and 80% negatives
static const ABWLParams wl_params_512[] = {
    {24, 18, 15, 17, 6, 13}, {19, 14, 13, 17, 2, 18}, {23, 19, 12, 15, 6, 19}, {24, 14, 16, 16, 6, 11},
    ...
};
```
- `beblid.p256.hpp` with similar content


Use them:
```
{
    #include "beblid.p512.hpp"
    #include "beblid.p256.hpp"
    if (n_bits == SIZE_512_BITS)
        wl_params_.assign(wl_params_512, wl_params_512 + sizeof(wl_params_512) / sizeof(wl_params_512[0]));
    else if(n_bits == SIZE_256_BITS)
        wl_params_.assign(wl_params_256, wl_params_256 + sizeof(wl_params_256) / sizeof(wl_params_256[0]));
    else
        CV_Error(...);
}
```

Reply:
Done. I located the files in [`opencv_contrib/modules/xfeatures2d/src/beblid.p512.hpp`](https://github.com/opencv/opencv_contrib/pull/2774/commits/0f10bd3ff16c6eb285dda2dbabd36675de544d10#diff-78494b0c74baff117f619067e1a8637dac3686f54ea8101ca146f8e9e06fa90c) and [`opencv_contrib/modules/xfeatures2d/src/beblid.p256.hpp`](https://github.com/opencv/opencv_contrib/pull/2774/commits/0f10bd3ff16c6eb285dda2dbabd36675de544d10#diff-06223661d4e54622d84f21aee4c0ec65b339e072d1b6044c0b11ea1c2e75bf1e) please let me known if this is the correct place.

Reply:
Location is correct.

Please remove `#ifndef __OPENCV_XFEATURES2D_BEBLID_P256_HPP__` guards from them. These headers are not expected to be used as "standalone".

Reply:
Done! Thank you again.